src/libraries/Microsoft.Win32.SystemEvents/src/Microsoft/Win32/SystemEvents.cs

```diff
@@ -1075,42 +1072,53 @@ private static void RemoveEventHandler(object key, Delegate? value)
-                lock (s_procLockObject)
+                if (s_systemEvents is null)
+                {
+                    return;
+                }
+
+                // If we are using system events from another thread, request that it terminate
+                if (s_windowThread is not null)
                 {
-                    if (s_systemEvents != null)
-                    {
-                        // If we are using system events from another thread, request that it terminate
-                        if (s_windowThread != null)
-                        {
 #if DEBUG
-                            unsafe
-                            {
-                                int pid;
-                                int thread = Interop.User32.GetWindowThreadProcessId(s_systemEvents._windowHandle, &pid);
-                                Debug.Assert(thread != Interop.Kernel32.GetCurrentThreadId(), "Don't call Shutdown on the system events thread");
+                    unsafe
+                    {
+                        int pid;
+                        int thread = Interop.User32.GetWindowThreadProcessId(s_systemEvents._windowHandle, &pid);
+                        Debug.Assert(thread != Interop.Kernel32.GetCurrentThreadId(), "Don't call Shutdown on the system events thread");
 
-                            }
+                    }
 #endif
-                            // The handle could be valid, Zero or invalid depending on the state of the thread
-                            // that is processing the messages. We optimistically expect it to be valid to
-                            // notify the thread to shutdown. The Zero or invalid values should be present
-                            // only when the thread is already shutting down due to external factors.
-                            if (s_systemEvents._windowHandle != IntPtr.Zero)
-                            {
-                                Interop.User32.PostMessageW(s_systemEvents._windowHandle, Interop.User32.WM_QUIT, IntPtr.Zero, IntPtr.Zero);
-                                GC.KeepAlive(s_systemEvents);
-                            }
-
-                            s_windowThread.Join();
-                        }
-                        else
-                        {
-                            s_systemEvents.Dispose();
-                            s_systemEvents = null;
-                        }
+                    // The handle could be valid, Zero or invalid depending on the state of the thread
+                    // that is processing the messages. We optimistically expect it to be valid to
+                    // notify the thread to shutdown. The Zero or invalid values should be present
+                    // only when the thread is already shutting down due to external factors.
+                    if (s_systemEvents._windowHandle != IntPtr.Zero)
+                    {
+                        InvokeOnEventsThread(() => Interop.User32.PostQuitMessage(0));
```

Comment:
Now that we are not waiting for the window thread to finish, what's the likelihood of the process ending before the message gets a chance to be delivered?

Would it be better to delete the shutdown handling completely to make the behavior deterministic (ie the message is never delivered during vs. the message is sometime delivered in the current delta)?

Reply:
Good point. It would be difficult to determine when message would get delivered vs. not. I think it is better to be deterministic about the behavior here and to remove shutdown handling completely. I'll update the PR accordingly.